antlr_ast/ast.py

```diff
@@ -283,3 +176,505 @@ def reportContextSensitivity(
     ):
         return
         # raise Exception("TODO")
+
+
+# Parse Tree Visitor ----------------------------------------------------------
+# TODO: visitor inheritance not really needed, but indicates compatibility
+# TODO: make general node (Terminal) accessible in class property (.subclasses)?
+
+
+class BaseNode(AST):
+    """AST is subclassed so we can use Python ast module  visiting and walking on the custom AST"""
+
+    def __init__(self, children, field_references, label_references, ctx=None):
+        self.children = children
+
+        self._field_references = field_references
+        self.children_by_field = materialize(self._field_references, self.children)
+
+        self._label_references = label_references
+        self.children_by_label = materialize(self._label_references, self.children)
+
+        self._ctx = ctx
+
+    subclasses = {}
+
+    _fields = ()
+
+    # whether to descend for selection (greater descends into lower)
+    _priority = 2
+
+    # getattr: return None or raise for nonexistent attr
+    # in Transformer conditionals:
+    # - getattr(obj, attr, None) works with both
+    # - hasattr(obj, attr) if strict
+    # - obj.attr if not strict
+    _strict = False
+
+    @classmethod
+    def create(cls, ctx, children=None):
+        if children is None:
+            children = ctx.children
+
+        field_names = get_field_names(ctx)
+        children_by_field = get_field_references(ctx, field_names)
+
+        label_names = get_label_names(ctx)
+        children_by_label = get_field_references(ctx, label_names)
+
+        cls_name = type(ctx).__name__.split("Context")[0]
+        subclass = cls.get_node_cls(cls_name, tuple(field_names))
+        return subclass(children, children_by_field, children_by_label, ctx)
+
+    @classmethod
+    def get_node_cls(cls, cls_name, field_names=tuple()):
```

Comment:
I would make a registry class but keep the method name, since it's not just registering classes, but returning existing ones and providing a lookup by name to make instance checks work in the transformation step.